src/rules/rules_of_hooks.rs

```diff
@@ -0,0 +1,378 @@
+      | Node::ForOfStmt(_)
+      | Node::FnExpr(_)
+      | Node::ForStmt(_)
+      | Node::WhileStmt(_)
+      | Node::ExportDefaultExpr(_)
+      | Node::TryStmt(_) => {
+        let _ = self.parent_kind.pop();
+      }
+      _ => {}
+    }
+  }
+
+  fn call_expr(&mut self, node: &CallExpr, ctx: &mut Context) {
+    if is_hook_call(node) {
+      if self.parent_kind.is_empty() {
+        ctx.add_diagnostic_with_hint(
+          node.range(),
+          CODE,
+          DiagnosticKind::OutsideComponent.message(),
+          DiagnosticKind::OutsideComponent.hint(),
+        );
+      }
+
+      for item in self.parent_kind.iter().rev() {
+        match item {
+          ParentKind::Unknown => break,
+          ParentKind::Var(name) => {
+            if !is_hook_or_component_name(name) {
+              ctx.add_diagnostic_with_hint(
+                node.range(),
+                CODE,
+                DiagnosticKind::OutsideComponent.message(),
+                DiagnosticKind::OutsideComponent.hint(),
+              );
+            }
+
+            break;
+          }
+          ParentKind::Fn((name, cond_count)) => {
+            if *cond_count > 0 {
+              ctx.add_diagnostic_with_hint(
+                node.range(),
+                CODE,
+                DiagnosticKind::Conditionally.message(),
+                DiagnosticKind::Conditionally.hint(),
+              );
+            } else if !is_hook_or_component_name(name) {
+              ctx.add_diagnostic_with_hint(
+                node.range(),
+                CODE,
+                DiagnosticKind::OutsideComponent.message(),
+                DiagnosticKind::OutsideComponent.hint(),
+              );
+            }
+
+            break;
+          }
+          ParentKind::Loop => {
+            ctx.add_diagnostic_with_hint(
+              node.range(),
+              CODE,
+              DiagnosticKind::Loop.message(),
+              DiagnosticKind::Loop.hint(),
+            );
+          }
+          ParentKind::TryCatch => {
+            ctx.add_diagnostic_with_hint(
+              node.range(),
+              CODE,
+              DiagnosticKind::TryCatch.message(),
+              DiagnosticKind::TryCatch.hint(),
+            );
+            break;
+          }
+          ParentKind::Bin | ParentKind::Cond => {
+            ctx.add_diagnostic_with_hint(
+              node.range(),
+              CODE,
+              DiagnosticKind::Conditionally.message(),
+              DiagnosticKind::Conditionally.hint(),
+            );
+            break;
+          }
+          _ => {}
+        }
+      }
+    }
+  }
+}
+
+fn is_hook_call(call_expr: &CallExpr) -> bool {
+  if let Callee::Expr(Expr::Ident(id)) = call_expr.callee {
+    return id.sym().starts_with("use");
+  }
+
+  false
+}
+
+fn is_hook_or_component_name(name: &str) -> bool {
+  if name.starts_with("use") {
+    return true;
+  }
+
+  if let Some(ch) = name.chars().next() {
+    return ch.is_uppercase();
+  }
```

Comment:
Where is this hit? Could you add a small unit test for this function?

Reply:
This is hit in pretty much all bad variants. It tests if the enclosing function is a component.